### PR TITLE
Issue 3419: Mitigate HSTS fingerprinting

### PR DESCRIPTION
**Pull request by jumde**
fixes https://github.com/brave/brave-browser/issues/3419
## Description

HSTS super-cookies are used as a fingerprinting vector.  This change disallow's third parties from setting trackable security headers like "Strict-Transport-Security", "Expect-CT", "Public-Key-Pins", "Public-Key-Pins-Report-Only"

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests`) on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Test Plan:

1. Navigate to `https://avatars2.githubusercontent.com` and disable `Connections Encrypted`
2. Navigate to `https://jsfiddle.net` and disable `Connections Encrypted` (see: https://github.com/brave/brave-browser/issues/3424)
3. Clear Browsing Data 
4. Navigate to `https://jsfiddle.net/x5mqb807/` and verify that `Connections Encrypted` is disabled
5. Verify that content from `avatars2.githubusercontent.com` loads with a 301 redirect
6. Navigate to ` http://avatars2.githubusercontent.com`, reload ` http://avatars2.githubusercontent.com` (not https) and verify it loads with a 307 redirect
7. Reload `https://jsfiddle.net/x5mqb807/` in the first tab and verify the image is loaded with a 307 redirect.

Before navigating to `avatars2.githubusercontent.com`:

screen shot 2019-02-20 at 2 07 33 pm

After navigating to  `avatars2.githubusercontent.com`:

screen shot 2019-02-20 at 2 04 04 pm

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

### CODE REVIEW

**Comment by diracdeltas** [comment]
Note to reviewers: this should match the behavior in https://github.com/brave/browser-laptop/pull/13649/files, minus the behavior where the HSTS cache is cleared on startup.

File: browser/net/brave_network_delegate_base.cc [hunk 1]
```diff
@@ -124,6 +151,10 @@ int BraveNetworkDelegateBase::OnHeadersReceived(
     const net::HttpResponseHeaders* original_response_headers,
     scoped_refptr<net::HttpResponseHeaders>* override_response_headers,
     GURL* allowed_unsafe_redirect_url) {
+  brave::RemoveTrackableSecurityHeadersForThirdParty(request,
```
Comment (diracdeltas) [inline browser/net/brave_network_delegate_base.cc]: does this need to check `if (!request)` first?

File: browser/net/brave_network_delegate_base_unittest.cc [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+  scoped_refptr<HttpResponseHeaders> headers(
+    new HttpResponseHeaders(net::HttpUtil::AssembleRawHeaders(
+                                                     raw_headers.c_str(),
+                                                     raw_headers.size())));
+
+  RemoveTrackableSecurityHeadersForThirdParty(request.get(), nullptr, &headers);
+  for (auto* header : kTrackableSecurityHeaders) {
+    EXPECT_FALSE(headers->HasHeader(header));
+  }
```
Comment (diracdeltas) [inline browser/net/brave_network_delegate_base_unittest.cc]: i suggest adding some other headers to raw_headers and testing that they are present in the result here

File: browser/net/brave_network_delegate_base.cc (hunk 1 above)
Comment (jumde) [reply browser/net/brave_network_delegate_base.cc]: done!

File: browser/net/brave_network_delegate_base_unittest.cc (hunk 2 above)
Comment (jumde) [reply browser/net/brave_network_delegate_base_unittest.cc]: done!

File: browser/net/brave_network_delegate_base.cc [hunk 3]
```diff
@@ -40,6 +44,30 @@ content::WebContents* GetWebContentsFromProcessAndFrameId(int render_process_id,
...
+void RemoveTrackableSecurityHeadersForThirdParty(
+        URLRequest* request,
+        const net::HttpResponseHeaders* original_response_headers,
+        scoped_refptr<net::HttpResponseHeaders>* override_response_headers) {
+  if (!request || (!original_response_headers &&
+                   !override_response_headers->get())) {
+    return;
+  }
+  if (net::registry_controlled_domains::SameDomainOrHost(
+        request->url(), request->site_for_cookies(),
```
Comment (iefremov) [inline browser/net/brave_network_delegate_base.cc]: Please use `request->top_frame_origin()`

File: browser/net/brave_network_delegate_base.h [hunk 4]
```diff
@@ -20,6 +27,18 @@ namespace net {
 class URLRequest;
 }
 
+namespace brave {
+  const std::unordered_set<std::string> kTrackableSecurityHeaders = {
```
Comment (iefremov) [inline browser/net/brave_network_delegate_base.h]: Objects with non-trivial destructors are generally banned by CC (as well as `unordered_maps`), its better (but a bit wordy) to use `static const base::NoDestructor<base::flat_set<base::StringPiece>>`

For example https://cs.chromium.org/chromium/src/components/google/core/common/google_util.cc?q=google_util&sq=package:chromium&g=0&l=137

File: browser/net/brave_network_delegate_base_unittest.cc [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+                            "pin-sha256=\"cUPcTAZWKaASuYWhhBAkE3h2+soZS7sWs=\";"
+                            "max-age=5184000; includeSubDomains\n"
+                            "Public-Key-Pins-Report-Only:"
+                            "pin-sha256=\"cUPcTAZWKaASukE3h2+soZS7sWs=\";"
+                            "includeSubDomains;"
+                            "report-uri=\"https://www.a.org/hpkp-report\"\n"
+                            "X-XSS-Protection: 0";
+
+  scoped_refptr<HttpResponseHeaders> headers(
+    new HttpResponseHeaders(net::HttpUtil::AssembleRawHeaders(
```
Comment (iefremov) [inline browser/net/brave_network_delegate_base_unittest.cc]: wrong indentation, please `clang-format` the whole file

**Comment by iefremov** [comment]
Had only a quick glance, will get back to the PR later today.

File: browser/net/brave_network_delegate_base.cc (hunk 3 above)
Comment (jumde) [reply browser/net/brave_network_delegate_base.cc]: Similar to `site_for_cookies`, `top_frame_origin` might not be set every-time. (see: https://cs.chromium.org/chromium/src/net/url_request/url_request.h?type=cs&q=top_frame_origin&sq=package:chromium&g=0&l=290)

I think using `GetTabURLFromRenderFrameInfo` might be better. Let me know what you think?

Comment (iefremov) [reply browser/net/brave_network_delegate_base.cc]: `site_for_cookies` is intentionally blank for nested 3rdparty frames while `top_frame_origin` is not. I think it may be empty for requests that are not related to any frame at all, so we can do nothing in this case.

Basically, the comment about its experimental nature is outdated, since Chromium had a successful experiment based on this value, and it will stay with us

File: browser/net/brave_network_delegate_base.h [hunk 6]
```diff
@@ -20,6 +28,17 @@ namespace net {
 class URLRequest;
 }
 
+namespace brave {
+static const base::NoDestructor<base::flat_set<base::StringPiece>>
```
Comment (iefremov) [inline browser/net/brave_network_delegate_base.h]: From `no_destructor.h`: "Furthermore, a NoDestructor<T> should never have global scope as that may require a static initializer"

So i'd suggest to make a function that returns this set, and make `NoDestructor` object a static variable in this function.

File: browser/net/brave_network_delegate_base.h [hunk 7]
```diff
@@ -20,6 +28,17 @@ namespace net {
 class URLRequest;
 }
 
+namespace brave {
```
Comment (iefremov) [inline browser/net/brave_network_delegate_base.h]: nit: using namespace here looks inconsistent with other code in this file. Don't really know if it is better to not use it here.

File: browser/net/brave_network_delegate_base_unittest.cc [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+TEST_F(BraveNetworkDelegateBaseTest, RemoveTrackableSecurityHeadersMixedCase) {
+  net::TestDelegate test_delegate;
+  GURL request_url(kThirdPartyDomain);
+  GURL tab_url(kFirstPartyDomain);
+  std::unique_ptr<net::URLRequest> request = context()->CreateRequest(
+      request_url, net::IDLE, &test_delegate, TRAFFIC_ANNOTATION_FOR_TESTS);
+
+  request->set_top_frame_origin(url::Origin::Create(tab_url));
+  std::string raw_headers =
```
Comment (iefremov) [inline browser/net/brave_network_delegate_base_unittest.cc]: why not make a constant common for all tests?

**Comment by iefremov** [comment]
I would also suggest making a solid browser test, though I understand that this may be time consuming.

**Comment by iefremov** [comment]
LGTM with nits.

File: browser/net/brave_network_delegate_base_browsertest.cc [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+  GURL third_party =
+      embedded_test_server()->GetURL("c.com", "/iframe_hsts.html");
+
+  RedirectObserver redirect_observer(active_contents());
+  ui_test_utils::NavigateToURL(browser(), third_party);
+
+  EXPECT_TRUE(redirect_observer.has_sts_header(third_party));
+}
+
+IN_PROC_BROWSER_TEST_F(BraveNetworkDelegateBaseBrowserTest, ThirdPartySTS) {
```
Comment (iefremov) [inline browser/net/brave_network_delegate_base_browsertest.cc]: I would also add a test checking that nothing is blocked if tracking protection is off

File: browser/net/brave_network_delegate_base_browsertest.cc [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+          std::pair<GURL, bool>(handle->GetURL(), has_sts_header));
+    }
+  }
+
+  bool has_sts_header(GURL url) const {
+    auto iter = sts_header_for_url.find(url);
+    if (iter != sts_header_for_url.end()) {
+      return iter->second;
+    }
+    return false;
```
Comment (iefremov) [inline browser/net/brave_network_delegate_base_browsertest.cc]: I think we can even DCHECK that something is found, otherwise the test can pass if there is a typo in subframe address, for example.

File: browser/net/brave_network_delegate_base_browsertest.cc [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+ private:
+  std::map<GURL, bool> sts_header_for_url;
```
Comment (iefremov) [inline browser/net/brave_network_delegate_base_browsertest.cc]: `sts_header_for_url_`

File: browser/net/brave_network_delegate_base_browsertest.cc [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+ public:
+  explicit RedirectObserver(WebContents* web_contents)
+      : WebContentsObserver(web_contents) {}
+  ~RedirectObserver() override = default;
+
+  void DidFinishNavigation(NavigationHandle* handle) override {
+    const net::HttpResponseHeaders* response = handle->GetResponseHeaders();
+    if (response) {
+      bool has_sts_header = response->HasHeader("Strict-Transport-Security");
+      sts_header_for_url.insert(
```
Comment (iefremov) [inline browser/net/brave_network_delegate_base_browsertest.cc]: just `sts_header_for_url[handle->GetURL()] = has_sts_header;`
or `sts_header_for_url.insert({handle->GetURL(), has_std_header});`

File: browser/net/brave_network_delegate_base_browsertest.cc [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+class RedirectObserver : public WebContentsObserver {
+ public:
+  explicit RedirectObserver(WebContents* web_contents)
+      : WebContentsObserver(web_contents) {}
+  ~RedirectObserver() override = default;
+
+  void DidFinishNavigation(NavigationHandle* handle) override {
+    const net::HttpResponseHeaders* response = handle->GetResponseHeaders();
+    if (response) {
+      bool has_sts_header = response->HasHeader("Strict-Transport-Security");
```
Comment (iefremov) [inline browser/net/brave_network_delegate_base_browsertest.cc]: nit: `const has_sts_header`, `const  GURL third_party`, etc

File: test/BUILD.gn [hunk 14]
```diff
@@ -283,6 +284,7 @@ test("brave_browser_tests") {
     "//brave/browser/extensions/brave_extension_functional_test.h",
     "//brave/browser/extensions/api/brave_shields_api_browsertest.cc",
     "//brave/browser/extensions/api/brave_theme_api_browsertest.cc",
+    "//brave/browser/net/brave_network_delegate_base_browsertest.cc",
```
Comment (iefremov) [inline test/BUILD.gn]: maybe merge this and the next one? Or give more descriptive names to source files: this one against hsts, etc
